### DISCUSSION TITLE
Scalability and User Management in boardgame.io

### DISCUSSION THREAD

**Discussion by matimusss** [boardgameio/boardgame.io#1173] [General]
Hi everyone! I'm Matias from Argentina. I've been using boardgame.io for a project and I'm really happy with the framework, so firstly thanks to the developers and the community. Currently, I'm triying to readabout scalability, user management, and server requirements such as bandwidth.

I'd appreciate guidance on best practices and key considerations for handling scalability in boardgame.io games. Any recommendations for managing multiple simultaneous users and server requirements?

I'm considering setting up an external 'lobby' to connect and communicate with boardgame.io servers based on user load, aiming for a horizontal scaling approach. I plan to create endpoints between the external lobby and bgio servers, then storing necessary data outisde BGIO servers.

Though I know certainty may come after testing, I'd like to have a good estimation. Any insights, experiences, or guidance on this would be highly appreciated. Thanks a lot!

(maybe max users, sv specs , processor, ram, bandwidth)

**Comment by fladrif** [comment]
What kind of scale are you thinking about? I personally don't have any experience with large scale usage on the framework, and would invite anyone who does to chime in, but in general the resource usage is generally low from my experience. 

On Scaling: In general it's not something to think about at early stages of any project. I think it's often considered a premature optimization problem. Unless you're trying to solve the problem just as an exercise, any amount of work you put in will most likely be wasted because no one can tell where your scaling bottlenecks are this early in the process.

General thoughts: Having a project currently running on boardgame.io, I would suggest trimming the state object as much as possible. Smaller game state objects are easier to debug and will save you storage in your backend (if you're planning on doing this) because the state object is duplicated for every move in the log (it holds the entire state in each iteraiton instead of just diffs).
